### PR TITLE
[master] Unreserve a FCP device after undeidcate a FCP device

### PR DESCRIPTION
**Pull request by changzhi1990**
In the current implementation, the FCP device will be unreserved firstly before executing the smcli undedicate command. 

But in multi-thread case, if one FCP device is set unreserved, maybe other thread will use the same FCP device and then will raise an error.

### CODE REVIEW

**Comment by dongyanyang** [comment]
Pls always remember to add some description in the PR to give other reviewers more background and also to keep a clear commit log.

File: zvmsdk/volumeop.py [hunk 1]
```diff
@@ -680,10 +680,6 @@ def _detach(self, fcp, assigner_id, target_wwpns, target_lun,
         """Detach a volume from a guest"""
         LOG.info('Start to detach device from %s' % assigner_id)
         connections = self.fcp_mgr.decrease_fcp_usage(fcp, assigner_id)
-        # Unreserved fcp device
-        fcp_connections = self.db.get_connections_from_fcp(fcp)
-        if not fcp_connections:
-            self.fcp_mgr.unreserve_fcp(fcp)
         if is_root_volume:
             LOG.info('Detaching device from %s is done.' % assigner_id)
```
Comment (dongyanyang) [inline zvmsdk/volumeop.py:684]: not that simple. 
if is_root_volume: 
 return directly without unreserve?

**Comment by dongyanyang** [comment]
> In the current implementation, the FCP device will be unreserved after executing smcli command even if the smcli command fails.
> 
> So in this pr, after executing the smcli command successfully and then unreserve the FCP device. And if zvmsdk executes smcli command fails, do not unreserve the FCP device.
> 
> Also, for the root volume, it needs to unreserve the FCP device.

This is not the intention for this PR...
This is to make sure the FCP cann't be allocated to other vm before the undedicate finishes, so that to avoid error in multi-thread case.

File: zvmsdk/volumeop.py (hunk 1 above)
Comment (dongyanyang) [reply zvmsdk/volumeop.py:692]: if unreserve here for root volume, the volume will be marked as free in DB, but the fcp will only be released after the vm is deleted. same error exists in multi thread case.
pls test whether the fcp will be left unreserved if we remove this unreserve for root volume. If not influence function, we can remove the unreserve.

**Comment by dongyanyang** [comment]
and pls always remember to paste the pep8 & py27 result when you make a new commit.

**Comment by changzhi1990** [comment]
py27 pass:
Ran 640 tests in 1.165s

OK (skipped=3)
py27 run-test: commands[1] | python -m unittest discover -v -s /root/work/python-zvm-sdk/smtLayer/tests/unit
test_getVM_directory_py3 (test_vmUtils.SMTvmUtilsTestCase) ... ok
test_getDiskPoolSpace3390 (test_getHost.SMTGetHostTestCase) ... ok
test_getDiskPoolSpace3390Unknown (test_getHost.SMTGetHostTestCase) ... ok
test_getDiskPoolSpace3390UnknownWithFlag (test_getHost.SMTGetHostTestCase) ... ok
test_getDiskPoolSpace9336 (test_getHost.SMTGetHostTestCase) ... ok
test_getDiskPoolSpace9336Unknown (test_getHost.SMTGetHostTestCase) ... ok
test_getDiskPoolSpace9336UnknownWithFlag (test_getHost.SMTGetHostTestCase) ... ok
test_getReservedMemSize (test_makeVM.SMTMakeVMTestCase) ... ok
test_getReservedMemSize_equal_size (test_makeVM.SMTMakeVMTestCase) ... ok
test_getReservedMemSize_gap_G (test_makeVM.SMTMakeVMTestCase) ... ok
test_getReservedMemSize_invalid_suffix (test_makeVM.SMTMakeVMTestCase) ... ok
test_getReservedMemSize_max_less_than_initial (test_makeVM.SMTMakeVMTestCase) ... ok

----------------------------------------------------------------------
Ran 12 tests in 0.002s

OK
_______________________________________________________ summary ________________________________________________________
  py27: commands succeeded
  congratulations :)


pep8 pass:
[root@xyimgr77 python-zvm-sdk]# tox -e pep8
pep8 develop-inst-noop: /root/work/python-zvm-sdk
pep8 installed: DEPRECATION: Python 2.7 reached the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 is no longer maintained. A future version of pip will drop support for Python 2.7. More details about Python 2 support in pip, can be found at https://pip.pypa.io/en/latest/development/release-process/#python-2-support,attrs==19.3.0,certifi==2020.4.5.2,chardet==3.0.4,configparser==4.0.2,contextlib2==0.6.0.post1,enum34==1.1.10,flake8==3.8.3,functools32==3.2.3.post2,idna==2.9,importlib-metadata==1.6.1,jsonschema==3.2.0,mccabe==0.6.1,netaddr==0.7.19,pathlib2==2.3.5,pycodestyle==2.6.0,pyflakes==2.2.0,PyJWT==1.7.1,pyrsistent==0.16.0,repoze.lru==0.7,requests==2.23.0,Routes==2.4.1,scandir==1.10.0,six==1.15.0,typing==3.7.4.1,urllib3==1.25.9,WebOb==1.8.6,zipp==1.2.0,-e git+https://github.com/openmainframeproject/python-zvm-sdk.git@456549f6a258aef0e36ccb2e8776bcfb0934c985#egg=zVMCloudConnector
pep8 run-test-pre: PYTHONHASHSEED='2148823066'
pep8 run-test: commands[0] | flake8
pep8 run-test: commands[1] | flake8 /root/work/python-zvm-sdk/scripts/sdkserver
_______________________________________________________ summary ________________________________________________________
  pep8: commands succeeded
  congratulations :)

File: zvmsdk/volumeop.py (hunk 1 above)
Comment (changzhi1990) [reply zvmsdk/volumeop.py:692]: I have tested this scenario. Remove this unreserve function for the root volume and then delete the vm. The fcp devices are unreserved after the vm is deleted.

File: zvmsdk/volumeop.py [hunk 2]
```diff
@@ -675,15 +675,18 @@ def _remove_disk(self, fcp, assigner_id, target_wwpns, target_lun,
...
+            LOG.info("Unreserve fcp device %s from detach", fcp)
+            self.fcp_mgr.unreserve_fcp(fcp)
+
     def _detach(self, fcp, assigner_id, target_wwpns, target_lun,
                 multipath, os_version, mount_point, is_root_volume):
         """Detach a volume from a guest"""
         LOG.info('Start to detach device from %s' % assigner_id)
         connections = self.fcp_mgr.decrease_fcp_usage(fcp, assigner_id)
-        # Unreserved fcp device
-        fcp_connections = self.db.get_connections_from_fcp(fcp)
```
Comment (dongyanyang) [inline zvmsdk/volumeop.py:684]: why query DB again instead of the previous returned connections directly?

File: zvmsdk/volumeop.py [hunk 3]
```diff
@@ -675,15 +675,18 @@ def _remove_disk(self, fcp, assigner_id, target_wwpns, target_lun,
                                       target_lun, multipath, os_version,
                                       mount_point, connections)
 
+    def _unreserve_fcp(self, fcp):
```
Comment (dongyanyang) [inline zvmsdk/volumeop.py:678]: if reuse the connections from caller function, then no need to make a seperate function

File: zvmsdk/volumeop.py [hunk 4]
```diff
@@ -680,10 +680,6 @@ def _detach(self, fcp, assigner_id, target_wwpns, target_lun,
         """Detach a volume from a guest"""
         LOG.info('Start to detach device from %s' % assigner_id)
         connections = self.fcp_mgr.decrease_fcp_usage(fcp, assigner_id)
```
Comment (jackydalong) [inline zvmsdk/volumeop.py:682]: Suggest we do remove_disk() first -> then do decrease_fcp_usage().
It's better to operate db after the real transaction completed successfully
@dongyanyang @changzhi1990

Comment (jackydalong) [reply zvmsdk/volumeop.py:682]: @jackydalong

**Pull request merged**